pym/bob/cmds/graph.py

```diff
@@ -0,0 +1,379 @@
+            if (tspan.node().getComputedTextLength() > 125) {
+                line.pop();
+                tspan.text(line.join(" "));
+                line = [word];
+                tspan = text.append("tspan").attr("dx", -120).attr("dy", lineHeight + dy + "em").text(word);
+            }
+        }
+    });
+}
+
+simulation.nodes(nodes).on('end', () => {
+  nodeElements
+    .attr('x', function (node) { return node.x })
+    .attr('y', function (node) { return node.y })
+  textElements
+    .attr('x', function (node) { return node.x })
+    .attr('y', function (node) { return node.y })
+  linkElements
+    .attr('x1', function (link) {
+        if (link.source.label.length > 21) {
+            return link.source.x + 65
+        } else {
+            return link.source.x + link.source.label.length * 3
+        }
+    })
+    .attr('y1', function (link) {
+            return link.source.y + ((link.source.label.length + 20) / 21-1) * 12
+        })
+    .attr('x2', function (link) {
+        if (link.target.label.length > 21) {
+            return link.target.x + 65
+        } else {
+            return link.target.x + link.target.label.length * 3 }
+        })
+    .attr('y2', function (link) {
+            return link.target.y + ((link.target.label.length + 20)/21-1) * 12
+        })
+})
+
+simulation.force("link").links(links)
+</script>
+""")
+
+def makeDotGraph(package, nodes, links):
+    with open("_".join(package.getStack()).replace(':','_') + ".dot", "w") as f:
+        f.write("digraph G {\n")
+        for (source,dest) in links:
+            f.write('"{}" -> "{}";\n'.format(source, dest))
+        f.write("}\n")
+
+def doGraph(argv, bobRoot):
+    parser = argparse.ArgumentParser(prog="bob graph",
+            description='Generate dependency graph')
+    parser.add_argument('package', nargs='?', help="Sub-package that is the root of the graph")
+
+    parser.add_argument('-D', default=[], action='append', dest="defines",
+        help="Override default environment variable")
+    parser.add_argument('-c', dest="configFile", default=[], action='append',
+        help="Use config File")
+    parser.add_argument('-e', '--exclude', default=[], action='append', dest="excludes",
+        help="Do not show this Package. (And all it's deps)")
+    parser.add_argument('-n', '--max-depth', type=int, default=-1, help="Max depth.")
+    parser.add_argument('-t', '--type', default='d3', help="Graph type. d3 (default) or dot")
+    args = parser.parse_args(argv)
+
+    defines = {}
+    for define in args.defines:
+        d = define.split("=")
+        if len(d) == 1:
+            defines[d[0]] = ""
+        elif len(d) == 2:
+            defines[d[0]] = d[1]
+        else:
+            parser.error("Malformed define: "+define)
+
+    recipes = RecipeSet()
+    recipes.setConfigFiles(args.configFile)
+    recipes.parse()
+    packages = recipes.generatePackages(lambda s,m: "unused", defines)
+    package = packages.walkPackagePath(args.package)
```

Comment:
It would be nice to support the query syntax too via `queryPackagePath`. The legacy `walkPackagePath` should be removed in the future...